### DISCUSSION TITLE
mscv fatal error LNK1181: cannot open input file

### DISCUSSION THREAD

**Discussion by andy5995** [mesonbuild/meson#13892] [Q&A] [answered]
```
The Meson build system
Version: 1.6.0
Source dir: D:\a\canfigger\canfigger
Build dir: D:\a\canfigger\canfigger\builddir
Build type: native build
Project name: canfigger
Project version: 0.3.1999
C compiler for the host machine: cl (msvc 19.41.34123 "Microsoft (R) C/C++ Optimizing Compiler Version 19.41.34123 for x64")
C linker for the host machine: link link 14.41.34123.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
```

When I try to build the tests:

fatal error LNK1181: cannot open input file 'canfigger.lib'

To build the library and declare it, I have

```
buildtarget = library(
  meson.project_name(),
  src,
  version : meson.project_version(),
  install: not meson.is_subproject()
  )

# How to use in a superproject and other info
# https://mesonbuild.com/Subprojects.html
canfigger_dep = declare_dependency(
  include_directories : include_directories('.'),
  link_with : buildtarget
  )
```

For the exe tests, I use  'canfigger_dep' as the dependency.

What is the proper way to do this so the tests will build on Windows? They already build on Linux, MacOS, and BSD. (All my changes are in [this pull request](https://github.com/andy5995/canfigger/pull/32/files).

**Comment by QuLogic** [comment, marked as answer]
MSVC defaults to hidden visibility; if no symbols are public in a shared library, it will not create the corresponding `.lib` file. You would need to tag the public symbols with `__declspec(dllexport)` or provide a symbol list with the `vs_module_defs` argument.

**Comment by andy5995** [comment]
Thanks @QuLogic I'm now using this def file:

```
; canfigger.def
LIBRARY "canfigger"
EXPORTS
    ; Export functions for use in other applications
    canfigger_parse_file
    canfigger_free_current_key_node_advance
    canfigger_free_current_attr_str_advance
```

The tests are building but failing to run:

exit status 3221225781 or signal 3221225653 SIGinvalid

[job-logs.txt](https://github.com/user-attachments/files/17711818/job-logs.txt)

**Comment by andy5995** [comment]
Seems to work better when I link statically:

```
if build_machine.system() != 'windows'
  buildtarget = library(
    meson.project_name(),
    src,
    version : meson.project_version(),
    install: not meson.is_subproject(),
    )
else
  buildtarget = static_library(
    meson.project_name(),
    src,
    install: not meson.is_subproject(),
    # vs_module_defs: 'canfigger.def'
    )
endif
```

Unfortunately, 1/8 tests passed, 1 failed, and the rest timed out.

**Reply by andy5995** [reply]
> Unfortunately, 1/8 tests passed, 1 failed, and the rest timed out.

Using "rb" as the second argument to fopen() fixed most. 6/8 pass now, 2 timeout. So... not really a meson issue of course.

So I guess the question remains, Should I just use static libraries on Windows? Or what step did I miss after I added the def file and all the tests built, I just got `exit status 3221225781 or signal 3221225653 SIGinvalid is:closed` for all of them?
